hive/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java

```diff
@@ -0,0 +1,211 @@
+    tableLocation = temp.newFolder("hive-").getPath();
+  }
+
+  @After
+  public void cleanup() {
+    catalog.dropTable(TABLE_IDENTIFIER);
+  }
+
+  @Test
+  public void testCreateTableTxn() {
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    txn.updateProperties()
+        .set("prop", "value")
+        .commit();
+
+    // verify the table is still not visible before the transaction is committed
+    Assert.assertFalse(catalog.tableExists(TABLE_IDENTIFIER));
+
+    txn.commitTransaction();
+
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    Assert.assertEquals("Table props should match", "value", table.properties().get("prop"));
+  }
+
+  @Test
+  public void testCreateTableTxnTableCreatedConcurrently() {
+    exceptionRule.expect(RuntimeException.class);
+    exceptionRule.expectMessage("Metastore operation failed");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+
+    // create the table concurrently
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testCreateTableTxnTableAlreadyExists() {
+    exceptionRule.expect(AlreadyExistsException.class);
+    exceptionRule.expectMessage("Table already exists: hivedb.tbl");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // create a table before starting a transaction
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testReplaceTableTxn() {
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    Assert.assertTrue("Table should exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newReplaceTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, PartitionSpec.unpartitioned(), tableLocation, Maps.newHashMap());
+    txn.commitTransaction();
+
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    Assert.assertEquals("Partition spec should match", PartitionSpec.unpartitioned(), table.spec());
+  }
+
+  @Test
+  public void testReplaceTableTxnTableNotExists() {
+    exceptionRule.expect(NoSuchTableException.class);
+    exceptionRule.expectMessage("No such table: hivedb.tbl");
```

Comment:
I would expect a replace to be idempotent and actually implement create or replace. What do you think?

Reply:
I interpret the API in Spark as follows:

**Case 1 (create)**
1. if exists -> fail
2. if created concurrently -> fail

**Case 2 (replace)**
1. if exists -> replace
2. if not exists -> fail
3. if deleted/modified concurrently -> up to the catalog

**Case 3 (create or replace)**
1. if exists -> replace
2. if not exists -> create
3. if deleted/modified concurrently -> up to the catalog

It seems appropriate to handle 2.3 and 3.3 differently. For example, I would expect 2.3 to rather fail than succeed. However, it would be more appropriate for 3.3 to succeed. So, exposing a flag and modifying logic accordingly should be reasonable. What's your opinion?

Reply:
In the Spark API, we added a `orCreate` flag to control this. I think that's probably the right call here because I agree with most of your descriptions of the expectations. I think that the behavior of cases 2.3 and 3.3 should _not_ be up to the catalog and it should instead match the behavior of the other cases: replace should fail if the table doesn't exist, and I would always expect that behavior, even if the table did exist at an earlier check.

Reply:
I think comments in [StagingTableCatalog](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/StagingTableCatalog.java) are a bit misleading w.r.t 2.3 and 3.3.

Reply:
We will have to modify `BaseTransaction` in order to support 2.3 and 3.3 as discussed.

One way to achieve that is to accept `TransactionType` in the constructor and maybe extend `TransactionType` with `CREATE_OR_REPLACE` as a separate transaction type. In the catalog, we will still have a boolean flag `orCreate`, but I think it would be easier to have a separate transaction type in `BaseTransaction`.

Reply:
Adding `CREATE_OR_REPLACE` sounds good to me.